pkg/nsx/services/expand.go

```diff
@@ -0,0 +1,317 @@
+	return &ipSetGroup
+}
+
+// Different direction rule decides different target of the traffic, we should carefully get
+// the destination pod selector and namespaces. Named port only cares about the destination
+// pod selector, so we use policy's AppliedTo or rule's AppliedTo when "IN" direction and
+// rule's DestinationSelector when "OUT" direction.
+func (service *SecurityPolicyService) getPodSelectors(obj *v1alpha1.SecurityPolicy, rule *v1alpha1.SecurityPolicyRule) ([]client.ListOptions, error) {
+	// Port means the target of traffic, so we should select the pod by rule direction,
+	// as for IN direction, we judge by rule's or policy's AppliedTo,
+	// as for OUT direction, then by rule's destinations.
+	// LabelSelect may return multiple namespaces
+	var selectorsUltimate []client.ListOptions
+	ruleDirection, err := getRuleDirection(rule)
+	if err != nil {
+		return nil, err
+	}
+
+	if ruleDirection == "IN" {
+		if len(obj.Spec.AppliedTo) > 0 {
+			for _, target := range obj.Spec.AppliedTo {
+				selector := client.ListOptions{}
+				if target.PodSelector != nil {
+					label, err := meta1.LabelSelectorAsSelector(target.PodSelector)
+					if err != nil {
+						return nil, err
+					}
+					selector.LabelSelector = label
+					selector.Namespace = obj.Namespace
+					selectorsUltimate = append(selectorsUltimate, selector)
+				}
+			}
+		} else if len(rule.AppliedTo) > 0 {
+			for _, target := range rule.AppliedTo {
+				// We only consider named port for PodSelector, not VMSelector
+				selector := client.ListOptions{}
+				if target.PodSelector != nil {
+					label, err := meta1.LabelSelectorAsSelector(target.PodSelector)
+					if err != nil {
+						return nil, err
+					}
+					selector.LabelSelector = label
+					selector.Namespace = obj.Namespace
+					selectorsUltimate = append(selectorsUltimate, selector)
+				}
+			}
+		}
+	} else if ruleDirection == "OUT" {
+		if len(rule.Destinations) > 0 {
+			var labelSelectors []client.ListOptions
+			var namespaceSelectors []client.ListOptions
+			for _, target := range rule.Destinations {
+				labelSelector := client.ListOptions{}
+				namespaceSelector := client.ListOptions{}
+				if target.PodSelector != nil {
+					label, err := meta1.LabelSelectorAsSelector(target.PodSelector)
+					if err != nil {
+						return nil, err
+					}
+					labelSelector.LabelSelector = label
+					labelSelectors = append(labelSelectors, labelSelector)
+				}
+				if target.NamespaceSelector != nil {
+					ns, err := service.ResolveNamespace(target.NamespaceSelector)
+					if err != nil {
+						return nil, err
+					}
+					for _, nsItem := range ns.Items {
+						namespaceSelector.Namespace = nsItem.Name
+						namespaceSelectors = append(namespaceSelectors, namespaceSelector)
+					}
+				} else {
+					namespaceSelector.Namespace = obj.Namespace
+					namespaceSelectors = append(namespaceSelectors, namespaceSelector)
+				}
+			}
+			if len(labelSelectors) > 0 && len(namespaceSelectors) > 0 {
```

Comment:
Considering this case:

apiVersion: apps/v1
kind: Deployment
metadata:
  name: mysql
  namespace: db2
  labels:
    role: db
spec:
  selector:
    matchLabels:
      role: db
  template:
    metadata:
      labels:
        role: db
    spec:
      containers:
        - image: "harbor-repo.vmware.com/dockerhub-proxy-cache/gophernet/netcat:latest"
          command: ["nc", "-l", "-p", "3306"]
          name: mysql2
          ports:
            - containerPort: 3306
              name: mysql-port

apiVersion: apps/v1
kind: Deployment
metadata:
  name: mysql
  namespace: db3
  labels:
    role: db3
spec:
  selector:
    matchLabels:
      role: db3
  template:
    metadata:
      labels:
        role: db3
    spec:
      containers:
        - image: "harbor-repo.vmware.com/dockerhub-proxy-cache/gophernet/netcat:latest"
          command: ["nc", "-l", "-p", "3306"]
          name: mysql3
          ports:
            - containerPort: 3306
              name: mysql-port

kubectl label ns db2 role=db
kubectl label ns db3 role=db3

we have rule like:
  rules:
    - direction: out
      action: allow
      appliedTo:
        - podSelector:
            matchLabels:
              role: web
      destinations:
        - podSelector:
            matchLabels:
              role: db
          namespaceSelector:
            matchLabels:
              role: db
        - namespaceSelector:
            matchLabels:
              role: db3
      ports:
        - protocol: TCP
          port: mysql-port

If we print at code line #L274
log.Info("getPodSelectors", "nsSelector", nsSelector.Namespace)
log.Info("getPodSelectors", "labelSelector", labelSelector.LabelSelector)
output:
2022-09-07 15:19:00.349	INFO	services/expand.go:280	getPodSelectors	{"nsSelector": "db2"}
2022-09-07 15:19:00.349	INFO	services/expand.go:281	getPodSelectors	{"labelSelector": "role=db"}
2022-09-07 15:19:00.349	INFO	services/expand.go:280	getPodSelectors	{"nsSelector": "db3"}
2022-09-07 15:19:00.349	INFO	services/expand.go:281	getPodSelectors	{"labelSelector": "role=db"}

You can see podSelector for "role=db" also applied to "nsSelector": "db3". I think this is not we want. We should select all pods that match namespaceSelector "nsSelector": "db3", that means all pods in namespace db3.  
As a result, we are not able to select pods from db3 , the result is showing only pod from db2: 172.20.0.74  is selected. 

db2                   mysql-c744b6ff8-9mwdl                   1/1     Running   0                2m41s   172.20.0.74   10.173.83.168   <none>           <none>
db3                   mysql-66c999f85d-ks4d7                  1/1     Running   0                2m41s   172.20.0.90   10.173.83.168   <none>           <none>

Therefore, I suggest we might need to consider how to fix it .

Reply:
fixed.